docs/source/ParticleNet.py

```diff
@@ -0,0 +1,289 @@
+      self.edge_convs.append(
+          EdgeConvBlock(k=k,
+                        in_feat=in_feat,
+                        out_feats=channels,
+                        cpu_mode=for_inference))
+
+    self.use_fusion = use_fusion
+    self.fusion_block = nn.Identity()
+    if self.use_fusion:
+      in_chn = sum(x[-1] for _, x in conv_params)
+      out_chn = int(np.clip((in_chn // 128) * 128, 128, 1024))
+      self.fusion_block = nn.Sequential(
+          nn.Conv1d(in_chn, out_chn, kernel_size=1, bias=False),
+          nn.BatchNorm1d(out_chn), nn.ReLU())
+
+    self.for_segmentation = for_segmentation
+
+    fcs = []
+    for idx, layer_param in enumerate(fc_params):
+      channels, drop_rate = layer_param
+      if idx == 0:
+        in_chn = out_chn if self.use_fusion else conv_params[-1][1][-1]
+      else:
+        in_chn = fc_params[idx - 1][0]
+      if self.for_segmentation:
+        fcs.append(
+            nn.Sequential(nn.Conv1d(in_chn, channels, kernel_size=1, bias=False),
+                          nn.BatchNorm1d(channels), nn.ReLU(),
+                          nn.Dropout(drop_rate)))
+      else:
+        fcs.append(
+            nn.Sequential(nn.Linear(in_chn, channels), nn.ReLU(),
+                          nn.Dropout(drop_rate)))
+    if self.for_segmentation:
+      fcs.append(nn.Conv1d(fc_params[-1][0], num_classes, kernel_size=1))
+    else:
+      fcs.append(nn.Linear(fc_params[-1][0], num_classes))
+    self.fc = nn.Sequential(*fcs)
+
+    self.for_inference = for_inference
+
+  def forward(self, points, features, mask):
+
+    points *= mask
+    features *= mask
+    coord_shift = (mask == 0) * 1e9
+    counts: Optional[torch.Tensor] = None
+    if self.use_counts:
+      counts = mask.float().sum(dim=-1)
+      counts = torch.max(counts, torch.ones_like(counts))  # >=1
+
+    fts = features
+    if self.bn_fts is not None and self.use_fts_bn:
+      fts = self.bn_fts(fts) * mask
+
+    outputs = []
+    for idx, conv in enumerate(self.edge_convs):
+      pts = (points if idx == 0 else fts) + coord_shift
+      fts = conv(pts, fts) * mask
+      if self.use_fusion:
+        outputs.append(fts)
+    if self.use_fusion:
+      fts = self.fusion_block(torch.cat(outputs, dim=1)) * mask
+
+    if self.for_segmentation:
+      x = fts
+    else:
+      if counts is not None and self.use_counts:
+        x = fts.sum(dim=-1) / counts  # divide by the real counts
+      else:
+        x = fts.mean(dim=-1)
+
+    output = self.fc(x)
+    if self.for_inference:
+      output = torch.softmax(output, dim=1)
+    return output
+
+
+class FeatureConv(nn.Module):
+  def __init__(self, in_chn, out_chn, **kwargs):
+    super(FeatureConv, self).__init__(**kwargs)
+    self.conv = nn.Sequential(
+        nn.BatchNorm1d(in_chn),
+        nn.Conv1d(in_chn, out_chn, kernel_size=1, bias=False),
+        nn.BatchNorm1d(out_chn), nn.ReLU())
+
+  def forward(self, x):
+    return self.conv(x)
+
+
+class ParticleNetTagger(nn.Module):
```

Comment:
We shouldn't actually need the ParticleNet code in coffea right? Just a jitted/traced serialized model?

Reply:
This might be the case, I'm still going over the examples to see the minimum files required for local testing. I'm having Claire push an example now so that I have something to work with and have this contribution be logged under her name.